{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/vagrant/miniconda/envs/dtailor27/bin/python\r\n"
     ]
    }
   ],
   "source": [
    "!which python"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# D-Tailor testing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Solution class\n",
    "- The sequence and all of its features stored in a single class\n",
    "- Troubleshooting\n",
    "    - ``Features.Structure.StructureMFE``?\n",
    "        - Runs ``Functions.analyze_structure_mfe``\n",
    "        - Which executes ``./3rdParty/unafold/ct-energy``\n",
    "    - ``Features.Structure.StructureDoubleStranded``?\n",
    "        - Runs ``Functions.analyze_structure_ds``\n",
    "        - Which executes ``perl 3rdParty/unafold/ss-count.pl``\n",
    "    - Need to troubleshoot and install properly all the third party programs...currently none working"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'cdsCAI': 0.6136121593930156,\n",
       " 'utrCdsStructureDoubleStranded': 'NA',\n",
       " 'utrCdsStructureDoubleStrandedList': 'NA',\n",
       " 'utrCdsStructureMFE': 0,\n",
       " 'utrNucleotideContentA': 0.45098039215686275,\n",
       " 'utrNucleotideContentAT': 0.6274509803921569,\n",
       " 'utrNucleotideContentC': 0.21568627450980393,\n",
       " 'utrNucleotideContentG': 0.1568627450980392,\n",
       " 'utrNucleotideContentGC': 0.37254901960784315,\n",
       " 'utrNucleotideContentT': 0.17647058823529413}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Solution import Solution\n",
    "from Features.CAI import CAI\n",
    "from Features.Structure import Structure, StructureMFE, StructureDoubleStranded\n",
    "from Features.NucleotideContent import NucleotideContent\n",
    "\n",
    "## Instantiate an object of type 'Solution'\n",
    "solution = Solution(sol_id='b0001', \n",
    "                    sequence='TATAGGCATAGCGCACAGACAGATAAAAATTACAGAGTACACAACATCCATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCACAGGTAACGGTGCGGGCTGA')\n",
    "\n",
    "\n",
    "## Instantiate Feature objects of interest\n",
    "\n",
    "# Feature to calculates the codon adaptation index\n",
    "cai_obj = CAI(solution=solution,label=\"cds\",args= {'cai_range':(49,115)})\n",
    "\n",
    "# Feature to predicts RNA Structure\n",
    "st1_obj = Structure(solution=solution,label=\"utrCds\",args= { 'structure_range' : (19,78) } )\n",
    "\n",
    "# Two sub-features inheriting from the class Structure\n",
    "st_mfe = StructureMFE(st1_obj)\n",
    "st_ss = StructureDoubleStranded(st1_obj)\n",
    "st1_obj.add_subfeature(st_mfe)\n",
    "st1_obj.add_subfeature(st_ss)\n",
    "\n",
    "# Feature to calculate nucleotide content\n",
    "nuc_obj = NucleotideContent(solution=solution ,label=\"utr\",args= { 'ntcontent_range':(0,50) } )\n",
    "\n",
    "\n",
    "## Add features to solution\n",
    "solution.add_feature(cai_obj)\n",
    "solution.add_feature(st1_obj)\n",
    "solution.add_feature(nuc_obj)\n",
    "\n",
    "\n",
    "## Retrieve feature score\n",
    "solution.scores\n",
    "# {'cdsCAI': 0.6136121593930156, 'utrCdsStructureDoubleStrandedList':[18, 19, 25, 26, 38, 39, 44, 45],\n",
    "# 'utrCdsStructureDoubleStranded': 8, 'utrCdsStructureMFE': -2.5,\n",
    "# 'utrNucleotideContentAT': 0.63, 'utrNucleotideContentG': 0.16,'utrNucleotideContentT' : 0.18,\n",
    "# 'utrNucleotideContentC': 0.22, 'utrNucleotideContentA' : 0.45, 'utrNucleotideContentGC': 0.37}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SequenceAnalyzer class\n",
    "- Lets you take in a bunch of sequences and compute the same properties for all of them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_name,sd_hyb_energy,mfe_structure,cai\n",
      "b0001 , NA , 0 , 0.613612159393\n",
      "b0002 , NA , 0 , 0.34043688741\n",
      "b0003 , NA , 0 , 0.341658034933\n",
      "b0004 , NA , 0 , 0.385891327353\n",
      "b0005 , NA , 0 , 0.377281853234\n",
      "b0006 , NA , 0 , 0.342733396212\n",
      "b0007 , NA , 0 , 0.319183029826\n",
      "b0008 , NA , 0 , 0.604195702312\n",
      "b0009 , NA , 0 , 0.396623675448\n",
      "b0010 , NA , 0 , 0.574062247682\n",
      "b0011 , NA , 0 , 0.286738246339\n",
      "b0013 , NA , 0 , 0.362374253526\n",
      "b0014 , NA , 0 , 0.723381361599\n",
      "b0015 , NA , 0 , 0.525547136369\n",
      "b0016 , NA , 0 , 0.224833425679\n",
      "b0018 , NA , 0 , 0.325397473007\n",
      "b4412 , NA , 0 , 0.388829489281\n",
      "b0019 , NA , 0 , 0.315147538129\n",
      "b0020 , NA , 0 , 0.326664585708\n",
      "b0021 , NA , 0 , 0.259970855677\n",
      "b0022 , NA , 0 , 0.35382053189\n",
      "b0023 , NA , 0 , 0.673378474366\n",
      "b0024 , NA , 0 , 0.223455416035\n",
      "b0025 , NA , 0 , 0.341974173688\n",
      "b0026 , NA , 0 , 0.544170808606\n",
      "b0027 , NA , 0 , 0.375930315257\n",
      "b0028 , NA , 0 , 0.446968612102\n",
      "b0029 , NA , 0 , 0.43576259638\n",
      "b0030 , NA , 0 , 0.33007758527\n",
      "b0031 , NA , 0 , 0.360583601428\n",
      "b0032 , NA , 0 , 0.419594043669\n",
      "b0033 , NA , 0 , 0.565238007841\n",
      "b0034 , NA , 0 , 0.274217763328\n",
      "b0035 , NA , 0 , 0.244827160564\n",
      "b0036 , NA , 0 , 0.403165554531\n",
      "b0037 , NA , 0 , 0.339820638368\n",
      "b0038 , NA , 0 , 0.452476250935\n",
      "b0039 , NA , 0 , 0.476208287595\n",
      "b0040 , NA , 0 , 0.389920755094\n",
      "b0041 , NA , 0 , 0.387332713602\n",
      "b0042 , NA , 0 , 0.467128281643\n",
      "b0043 , NA , 0 , 0.387986561249\n",
      "b0044 , NA , 0 , 0.398836859118\n",
      "b0045 , NA , 0 , 0.298070086981\n",
      "b0046 , NA , 0 , 0.358449938127\n",
      "b0047 , NA , 0 , 0.339327831857\n",
      "b0048 , NA , 0 , 0.400136382299\n",
      "b0049 , NA , 0 , 0.389342815555\n",
      "b0050 , NA , 0 , 0.29573399822\n",
      "b0051 , NA , 0 , 0.373683932955\n",
      "b0052 , NA , 0 , 0.307689655904\n"
     ]
    }
   ],
   "source": [
    "from SequenceAnalyzer import SequenceAnalyzer\n",
    "from Features import CAI,Structure,RNADuplex\n",
    "from Functions import validateCDS\n",
    "class TranslationFeaturesEcoliAnalyzer(SequenceAnalyzer):\n",
    "\n",
    "    '''\n",
    "    Class to analyze CAI, SD strength and structure in E. coli\n",
    "    '''\n",
    "\n",
    "    def __init__(self, input_file, input_type):\n",
    "        SequenceAnalyzer.__init__(self,input_file,input_type)\n",
    "\n",
    "    def configureSolution(self, solution):\n",
    "        solution.valid = validateCDS(solution.sequence[49:])\n",
    "        if solution.valid:\n",
    "            #CAI\n",
    "            cai_obj = CAI.CAI(solution=solution,\n",
    "                              label=\"cds\",\n",
    "                              args= { 'cai_range' :(49,len(solution.sequence)) } )\n",
    "\n",
    "            #Look for RBS\n",
    "            dup_obj1 = RNADuplex.RNADuplexRibosome(solution1=solution, \n",
    "                                                   label=\"sd16s\",\n",
    "                                                   args = { 'rnaMolecule1region' : (25,48) })\n",
    "            dup_mfe = RNADuplex.RNADuplexMFE(dup_obj1)\n",
    "            dup_obj1.add_subfeature(dup_mfe)\n",
    "\n",
    "            #MFE [-30,30]\n",
    "            st1_obj = Structure.Structure(solution=solution,\n",
    "                                          label=\"utr\", \n",
    "                                          args= { 'structure_range' : (49-30,49+30) } )\n",
    "            st_mfe = Structure.StructureMFE(st1_obj)\n",
    "            st1_obj.add_subfeature(st_mfe)\n",
    "\n",
    "            solution.add_feature(cai_obj)\n",
    "            solution.add_feature(dup_obj1)\n",
    "            solution.add_feature(st1_obj)\n",
    "    \n",
    "    def outputStart(self):\n",
    "        print \"gene_name,sd_hyb_energy,mfe_structure,cai\"\n",
    "\n",
    "    def output(self, solution):\n",
    "        if solution.valid:\n",
    "            print solution.solid,',',\n",
    "            print solution.scores['sd16sRNADuplexMFE'],',',\n",
    "            print solution.scores['utrStructureMFE'],',',\n",
    "            print solution.scores['cdsCAI']\n",
    "\n",
    "seqAnalyzerTest = TranslationFeaturesEcoliAnalyzer(\"testFiles/genomes/partial_ecoli_genome.csv\",\"CSV\")\n",
    "seqAnalyzerTest.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SequenceDesigner class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def configureSolution(self, solution):\n",
    "    '''\n",
    "    Solution configuration\n",
    "    '''\n",
    "    if solution.sequence == None:\n",
    "        return 0\n",
    "\n",
    "    ## Designer specific\n",
    "    solution.mutable_region=range(0,len(solution.sequence)) # whole region\n",
    "    solution.cds_region = (49,len(solution.sequence))\n",
    "    solution.keep_aa = True\n",
    "\n",
    "    ## Populate solution with desired features\n",
    "\n",
    "    # CAI\n",
    "    cai_obj = CAI.CAI(solution = solution,\n",
    "                      label=\"cds\",\n",
    "                      args = {'cai_range': (49,len(solution.sequence)),\n",
    "                              'mutable_region': range(49,len(solution.sequence)) } )\n",
    "\n",
    "    # Search SD\n",
    "    dup_obj1 = RNADuplex.RNADuplexRibosome(solution1=solution, label=\"sd16s\",\n",
    "                                            args = { 'rnaMolecule1region' : (25,48),\n",
    "                                            'mutable_region' : range(25,48) })\n",
    "    dup_mfe = RNADuplex.RNADuplexMFE(dup_obj1)\n",
    "    dup_obj1.add_subfeature(dup_mfe)\n",
    "\n",
    "    # MFE [-30,30]\n",
    "    st1_obj = Structure.Structure(solution=solution,label=\"utr\",\n",
    "                                args = {'structure_range' : (49-30,49+30), \n",
    "                                        'mutable_region' : range(49-30,49+30)} )\n",
    "    st_mfe = Structure.StructureMFE(st1_obj)\n",
    "    st1_obj.add_subfeature(st_mfe)\n",
    "\n",
    "    solution.add_feature(cai_obj)\n",
    "    solution.add_feature(dup_obj1)\n",
    "    solution.add_feature(st1_obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1.1.1',\n",
       " '1.1.3',\n",
       " '1.1.2',\n",
       " '1.1.5',\n",
       " '1.1.4',\n",
       " '1.3.1',\n",
       " '1.3.3',\n",
       " '1.3.2',\n",
       " '1.3.5',\n",
       " '1.3.4',\n",
       " '1.2.1',\n",
       " '1.2.3',\n",
       " '1.2.2',\n",
       " '1.2.5',\n",
       " '1.2.4',\n",
       " '1.5.1',\n",
       " '1.5.3',\n",
       " '1.5.2',\n",
       " '1.5.5',\n",
       " '1.5.4',\n",
       " '1.4.1',\n",
       " '1.4.3',\n",
       " '1.4.2',\n",
       " '1.4.5',\n",
       " '1.4.4',\n",
       " '3.1.1',\n",
       " '3.1.3',\n",
       " '3.1.2',\n",
       " '3.1.5',\n",
       " '3.1.4',\n",
       " '3.3.1',\n",
       " '3.3.3',\n",
       " '3.3.2',\n",
       " '3.3.5',\n",
       " '3.3.4',\n",
       " '3.2.1',\n",
       " '3.2.3',\n",
       " '3.2.2',\n",
       " '3.2.5',\n",
       " '3.2.4',\n",
       " '3.5.1',\n",
       " '3.5.3',\n",
       " '3.5.2',\n",
       " '3.5.5',\n",
       " '3.5.4',\n",
       " '3.4.1',\n",
       " '3.4.3',\n",
       " '3.4.2',\n",
       " '3.4.5',\n",
       " '3.4.4',\n",
       " '2.1.1',\n",
       " '2.1.3',\n",
       " '2.1.2',\n",
       " '2.1.5',\n",
       " '2.1.4',\n",
       " '2.3.1',\n",
       " '2.3.3',\n",
       " '2.3.2',\n",
       " '2.3.5',\n",
       " '2.3.4',\n",
       " '2.2.1',\n",
       " '2.2.3',\n",
       " '2.2.2',\n",
       " '2.2.5',\n",
       " '2.2.4',\n",
       " '2.5.1',\n",
       " '2.5.3',\n",
       " '2.5.2',\n",
       " '2.5.5',\n",
       " '2.5.4',\n",
       " '2.4.1',\n",
       " '2.4.3',\n",
       " '2.4.2',\n",
       " '2.4.5',\n",
       " '2.4.4',\n",
       " '5.1.1',\n",
       " '5.1.3',\n",
       " '5.1.2',\n",
       " '5.1.5',\n",
       " '5.1.4',\n",
       " '5.3.1',\n",
       " '5.3.3',\n",
       " '5.3.2',\n",
       " '5.3.5',\n",
       " '5.3.4',\n",
       " '5.2.1',\n",
       " '5.2.3',\n",
       " '5.2.2',\n",
       " '5.2.5',\n",
       " '5.2.4',\n",
       " '5.5.1',\n",
       " '5.5.3',\n",
       " '5.5.2',\n",
       " '5.5.5',\n",
       " '5.5.4',\n",
       " '5.4.1',\n",
       " '5.4.3',\n",
       " '5.4.2',\n",
       " '5.4.5',\n",
       " '5.4.4',\n",
       " '4.1.1',\n",
       " '4.1.3',\n",
       " '4.1.2',\n",
       " '4.1.5',\n",
       " '4.1.4',\n",
       " '4.3.1',\n",
       " '4.3.3',\n",
       " '4.3.2',\n",
       " '4.3.5',\n",
       " '4.3.4',\n",
       " '4.2.1',\n",
       " '4.2.3',\n",
       " '4.2.2',\n",
       " '4.2.5',\n",
       " '4.2.4',\n",
       " '4.5.1',\n",
       " '4.5.3',\n",
       " '4.5.2',\n",
       " '4.5.5',\n",
       " '4.5.4',\n",
       " '4.4.1',\n",
       " '4.4.3',\n",
       " '4.4.2',\n",
       " '4.4.5',\n",
       " '4.4.4']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from DesignOfExperiments.Design import FullFactorial\n",
    "\n",
    "#Design Methodology and thresholds\n",
    "design_param = { \"sd16sRNADuplexMFE\": {'type' : 'REAL' ,\n",
    "                                       'thresholds' : {'1': (-12.7,-7.3), '2': (-7.3,-5.8),\n",
    "                                                       '3': (-5.8,-5.2), '4': (-5.2,-3.3), '5': (-3.3, 2.0) } },\n",
    "                \"utrStructureMFE\": {'type' : 'REAL' ,\n",
    "                                    'thresholds' : {'1': (-29.2,-12.2), '2': (-12.2,-9.95),\n",
    "                                                    '3': (-9.95,-8.4), '4': (-8.4,-6.73), '5': (-6.73,0.65) } },\n",
    "                \"cdsCAI\" : {'type' : 'REAL' ,\n",
    "                            'thresholds' : {'1': (0.13,0.29), '2': (0.29,0.33),\n",
    "                                            '3': (0.33,0.37), '4': (0.37,0.42), '5': (0.42,0.86) } } }\n",
    "design = FullFactorial([\"sd16sRNADuplexMFE\",\"utrStructureMFE\",\"cdsCAI\"],design_param)\n",
    "design.listDesigns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "looking for combination:  5\n",
      "Solution found... inserting into DB...\n",
      "\n",
      "###########################\n",
      "# Optimized solution:\n",
      "# ID:  43437185966571386930924946858112457488\n",
      "# Sequence:  ttattaccggacaataatatttcaattcattaaagaggagaaaggtaccatggccctgtggatgcgccttctgcctctgctggcgctgctggccctgtgggggcctgacccagccgcagcctttgtgaaccagcacctgtgcggctcacacctggtagaagctctttacctagtttgcggtgaacgaggtttcttctacacaccgaagacccgtcgggaggctgaggacctgcaggttgggcaggtggagctgggtgggggccctggtgcaggctctctgcagcccttggccctggaggggtccctgcagaaacgtggcattgtagaacagtgctgtacctctatctgctctctctaccagctggagaactactgcaactag\n",
      "# Scores:  ['cdsCAI: 0.420465418221']\n",
      "# Levels:  ['cdsCAILevel: 5']\n",
      "# Number of generated solutions:  22\n",
      "# Distance to seed:  25\n",
      "###########################\n",
      "\n",
      "Program finished, all combinations were found...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(22, 25, 3.141724799297725)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from RunningExamples.Designer.TranslationFeaturesEcoliDesigner import TranslationFeaturesEcoliDesigner\n",
    "from DesignOfExperiments.Design import FullFactorial, Optimization\n",
    "\n",
    "#Seed sequence from which mutants will be derived\n",
    "seed='ttattaccggacaataatatttcaattcattaaagaggagaaaggtaccatggccctgtggatgcgcctcctgcccctgctggcgctgctggccctctggggacctgacccagccgcagcctttgtgaaccaacacctgtgcggctcacacctggtggaagctctctacctagtgtgcggggaacgaggcttcttctacacacccaagacccgccgggaggcagaggacctgcaggtggggcaggtggagctgggcgggggccctggtgcaggcagcctgcagcccttggccctggaggggtccctgcagaagcgtggcattgtggaacaatgctgtaccagcatctgctccctctaccagctggagaactactgcaactag'\n",
    "\n",
    "#Design Methodology and thresholds\n",
    "design_param = { \n",
    " \"cdsCAI\" : { 'type' : 'REAL' ,\n",
    " 'thresholds' : { '1': (0.13,0.29), '2': (0.29,0.33),\n",
    " '3': (0.33,0.37), '4': (0.37,0.42),\n",
    "'5': (0.42,0.86) } }\n",
    " }\n",
    "\n",
    "design = Optimization([\"cdsCAI\"],design_param, '5')\n",
    "\n",
    "output_file = \"/Users/nmih/Projects/codon_optimization/software_testing/D-Tailor/testFiles/outputFiles/tfec_1\"\n",
    "\n",
    "tirap_designer = TranslationFeaturesEcoliDesigner(\"tfec\", \n",
    "                                                  seed, \n",
    "                                                  design,\n",
    "                                                  output_file, \n",
    "                                                  createDB=True)\n",
    "tirap_designer.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'features': {'cdsCAI': {'thresholds': {'1': (0.13, 0.29),\n",
       "    '2': (0.29, 0.33),\n",
       "    '3': (0.33, 0.37),\n",
       "    '4': (0.37, 0.42),\n",
       "    '5': (0.42, 0.86)},\n",
       "   'type': 'REAL'}},\n",
       " 'featuresList': ['cdsCAI'],\n",
       " 'listDesigns': ['1', '3', '2', '5', '4'],\n",
       " 'nDesigns': 5,\n",
       " 'n_features': 1,\n",
       " 'thresholds': {'cdsCAI': {'1': (0.13, 0.29),\n",
       "   '2': (0.29, 0.33),\n",
       "   '3': (0.33, 0.37),\n",
       "   '4': (0.37, 0.42),\n",
       "   '5': (0.42, 0.86)}}}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "design.__dict__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Own code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating CAI for Pk? \n",
    "\n",
    "- Just use the simple Biopython calculator - but needed to modify code to get it to skip the unknown bases (N)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqUtils.CodonUsage import CodonAdaptationIndex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "illegal codon ACN in gene: lcl|NW_018150436.1_cds_XP_020546087.1_933\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon CGN in gene: lcl|NW_018150439.1_cds_XP_020545092.1_1667\n",
      "illegal codon TA in gene: lcl|NW_018150441.1_cds_XP_020544721.1_2022\n",
      "illegal codon GAN in gene: lcl|NW_018150443.1_cds_XP_020544345.1_2469\n",
      "illegal codon GAN in gene: lcl|NW_018150449.1_cds_XP_020543512.1_3267\n",
      "illegal codon AG in gene: lcl|NW_018150449.1_cds_XP_020543561.1_3371\n",
      "illegal codon GNC in gene: lcl|NW_018150463.1_cds_XP_020542560.1_4293\n",
      "illegal codon AA in gene: lcl|NW_018150484.1_cds_XP_020541991.1_4888\n",
      "illegal codon GA in gene: lcl|NW_018150525.1_cds_XP_020541789.1_5077\n",
      "illegal codon A in gene: lcl|NW_018150531.1_cds_XP_020541769.1_5099\n",
      "illegal codon AG in gene: lcl|NW_018150535.1_cds_XP_020541760.1_5107\n",
      "illegal codon AT in gene: lcl|NW_018150536.1_cds_XP_020541754.1_5111\n",
      "illegal codon A in gene: lcl|NW_018150539.1_cds_XP_020541749.1_5118\n",
      "illegal codon C in gene: lcl|NW_018150550.1_cds_XP_020541720.1_5148\n",
      "illegal codon G in gene: lcl|NW_018150603.1_cds_XP_020541604.1_5265\n",
      "illegal codon TA in gene: lcl|NW_018150630.1_cds_XP_020541556.1_5312\n",
      "illegal codon A in gene: lcl|NW_018150642.1_cds_XP_020541541.1_5328\n",
      "illegal codon TT in gene: lcl|NW_018150666.1_cds_XP_020541510.1_5359\n",
      "illegal codon A in gene: lcl|NW_018150689.1_cds_XP_020541486.1_5382\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'AAA': 1.0,\n",
       " 'AAC': 0.6835958956781579,\n",
       " 'AAG': 0.8501819970486965,\n",
       " 'AAT': 1.0,\n",
       " 'ACA': 0.9698236035279295,\n",
       " 'ACC': 0.6997060058798824,\n",
       " 'ACG': 0.3369172616547669,\n",
       " 'ACT': 1.0,\n",
       " 'AGA': 1.0,\n",
       " 'AGC': 0.33871546465733765,\n",
       " 'AGG': 0.4327671815167005,\n",
       " 'AGT': 0.6517581628991748,\n",
       " 'ATA': 0.44500644063390016,\n",
       " 'ATC': 0.560287878608823,\n",
       " 'ATG': 1.0,\n",
       " 'ATT': 1.0,\n",
       " 'CAA': 1.0,\n",
       " 'CAC': 0.4904431664411366,\n",
       " 'CAG': 0.5018068436200831,\n",
       " 'CAT': 1.0,\n",
       " 'CCA': 1.0,\n",
       " 'CCC': 0.27423029492297724,\n",
       " 'CCG': 0.26806429362707185,\n",
       " 'CCT': 0.6827540079008424,\n",
       " 'CGA': 0.13356785576870225,\n",
       " 'CGC': 0.059216914930650975,\n",
       " 'CGG': 0.11079352864654904,\n",
       " 'CGT': 0.2555012000952747,\n",
       " 'CTA': 0.45494302516671425,\n",
       " 'CTC': 0.2547365847695881,\n",
       " 'CTG': 0.26136440804574285,\n",
       " 'CTT': 0.4427739070203317,\n",
       " 'GAA': 1.0,\n",
       " 'GAC': 0.4646572115247604,\n",
       " 'GAG': 0.5341032773007619,\n",
       " 'GAT': 1.0,\n",
       " 'GCA': 1.0,\n",
       " 'GCC': 0.4917465037926556,\n",
       " 'GCG': 0.2468331796194169,\n",
       " 'GCT': 0.8223192606674321,\n",
       " 'GGA': 0.5128889736499158,\n",
       " 'GGC': 0.3511013635930199,\n",
       " 'GGG': 0.27271860398588726,\n",
       " 'GGT': 1.0,\n",
       " 'GTA': 0.2980785532636338,\n",
       " 'GTC': 0.45507205425261377,\n",
       " 'GTG': 0.3633088443063012,\n",
       " 'GTT': 1.0,\n",
       " 'TAA': 1.0,\n",
       " 'TAC': 0.6870831268818843,\n",
       " 'TAG': 0.6097455799913757,\n",
       " 'TAT': 1.0,\n",
       " 'TCA': 0.8974345174022247,\n",
       " 'TCC': 0.6188733405095085,\n",
       " 'TCG': 0.4076785073555795,\n",
       " 'TCT': 1.0,\n",
       " 'TGA': 0.6722725312634756,\n",
       " 'TGC': 0.41045642382875985,\n",
       " 'TGG': 1.0,\n",
       " 'TGT': 1.0,\n",
       " 'TTA': 0.8421410042238792,\n",
       " 'TTC': 0.649768556417164,\n",
       " 'TTG': 1.0,\n",
       " 'TTT': 1.0}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pk_cds = '/vagrant/gsl/notebooks/pk/GCF_001983325.1_ASM198332v1_cds_from_genomic.fna'\n",
    "\n",
    "biop_cai = CodonAdaptationIndex()\n",
    "biop_cai.generate_index(fasta_file=pk_cds)\n",
    "\n",
    "biop_cai.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "illegal codon A in gene: lcl|NC_000913.3_cds_261\n",
      "illegal codon AA in gene: lcl|NC_000913.3_cds_500\n",
      "illegal codon G in gene: lcl|NC_000913.3_cds_566\n",
      "illegal codon GA in gene: lcl|NC_000913.3_cds_632\n",
      "illegal codon AA in gene: lcl|NC_000913.3_cds_1011\n",
      "illegal codon GA in gene: lcl|NC_000913.3_cds_1165\n",
      "illegal codon GA in gene: lcl|NC_000913.3_cds_1170\n",
      "illegal codon GA in gene: lcl|NC_000913.3_cds_1428\n",
      "illegal codon T in gene: lcl|NC_000913.3_cds_1471\n",
      "illegal codon AA in gene: lcl|NC_000913.3_cds_1520\n",
      "illegal codon A in gene: lcl|NC_000913.3_cds_1523\n",
      "illegal codon A in gene: lcl|NC_000913.3_cds_1741\n",
      "illegal codon A in gene: lcl|NC_000913.3_cds_1957\n",
      "illegal codon AA in gene: lcl|NC_000913.3_cds_1985\n",
      "illegal codon AA in gene: lcl|NC_000913.3_cds_2107\n",
      "illegal codon GA in gene: lcl|NC_000913.3_cds_2153\n",
      "illegal codon AA in gene: lcl|NC_000913.3_cds_2648\n",
      "illegal codon GA in gene: lcl|NC_000913.3_cds_2650\n",
      "illegal codon AA in gene: lcl|NC_000913.3_cds_2682\n",
      "illegal codon A in gene: lcl|NC_000913.3_cds_2845\n",
      "illegal codon GA in gene: lcl|NC_000913.3_cds_2849\n",
      "illegal codon AG in gene: lcl|NC_000913.3_cds_2852\n",
      "illegal codon GA in gene: lcl|NC_000913.3_cds_2992\n",
      "illegal codon AA in gene: lcl|NC_000913.3_cds_3253\n",
      "illegal codon GA in gene: lcl|NC_000913.3_cds_3400\n",
      "illegal codon AA in gene: lcl|NC_000913.3_cds_3466\n",
      "illegal codon A in gene: lcl|NC_000913.3_cds_3746\n",
      "illegal codon GA in gene: lcl|NC_000913.3_cds_3754\n",
      "illegal codon GA in gene: lcl|NC_000913.3_cds_4059\n",
      "illegal codon AA in gene: lcl|NC_000913.3_cds_4226\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'AAA': 1.0,\n",
       " 'AAC': 1.0,\n",
       " 'AAG': 0.30580366868546033,\n",
       " 'AAT': 0.8192832417023009,\n",
       " 'ACA': 0.3024247937513724,\n",
       " 'ACC': 1.0,\n",
       " 'ACG': 0.6179930361680103,\n",
       " 'ACT': 0.38203833244455593,\n",
       " 'AGA': 0.09389264651984139,\n",
       " 'AGC': 1.0,\n",
       " 'AGG': 0.053710058974444405,\n",
       " 'AGT': 0.5461271465107783,\n",
       " 'ATA': 0.14351505798953537,\n",
       " 'ATC': 0.8276227907313191,\n",
       " 'ATG': 1.0,\n",
       " 'ATT': 1.0,\n",
       " 'CAA': 0.5324305784914001,\n",
       " 'CAC': 0.7513912549687678,\n",
       " 'CAG': 1.0,\n",
       " 'CAT': 1.0,\n",
       " 'CCA': 0.36413814855701,\n",
       " 'CCC': 0.23554644377858383,\n",
       " 'CCG': 1.0,\n",
       " 'CCT': 0.3010250749093203,\n",
       " 'CGA': 0.16276280278545963,\n",
       " 'CGC': 1.0,\n",
       " 'CGG': 0.2468263752373971,\n",
       " 'CGT': 0.95005497617699,\n",
       " 'CTA': 0.07402207456420806,\n",
       " 'CTC': 0.21016708832605854,\n",
       " 'CTG': 1.0,\n",
       " 'CTT': 0.20879090327782254,\n",
       " 'GAA': 1.0,\n",
       " 'GAC': 0.5945878054350308,\n",
       " 'GAG': 0.4502616829367878,\n",
       " 'GAT': 1.0,\n",
       " 'GCA': 0.5993556953483816,\n",
       " 'GCC': 0.7571450338477612,\n",
       " 'GCG': 1.0,\n",
       " 'GCT': 0.45262401776191197,\n",
       " 'GGA': 0.2670425042652622,\n",
       " 'GGC': 1.0,\n",
       " 'GGG': 0.3731424473950993,\n",
       " 'GGT': 0.8340874811463047,\n",
       " 'GTA': 0.41535146630848446,\n",
       " 'GTC': 0.5829695545108574,\n",
       " 'GTG': 1.0,\n",
       " 'GTT': 0.6976997985224983,\n",
       " 'TAA': 1.0,\n",
       " 'TAC': 0.7572797892154637,\n",
       " 'TAG': 0.1248695652173913,\n",
       " 'TAT': 1.0,\n",
       " 'TCA': 0.4466569236390209,\n",
       " 'TCC': 0.5368560467665328,\n",
       " 'TCG': 0.557590427475338,\n",
       " 'TCT': 0.5264431859700401,\n",
       " 'TGA': 0.49252173913043484,\n",
       " 'TGC': 1.0,\n",
       " 'TGG': 1.0,\n",
       " 'TGT': 0.7963025588997857,\n",
       " 'TTA': 0.2635185854485807,\n",
       " 'TTC': 0.742838305587605,\n",
       " 'TTG': 0.25937612944479965,\n",
       " 'TTT': 1.0}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pk_cds = '/vagrant/gsl/notebooks/ec/GCF_000005845.2_ASM584v2_cds_from_genomic.fna'\n",
    "\n",
    "biop_cai = CodonAdaptationIndex()\n",
    "biop_cai.generate_index(fasta_file=pk_cds)\n",
    "\n",
    "biop_cai.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:dtailor27]",
   "language": "python",
   "name": "conda-env-dtailor27-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
