{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/vagrant/miniconda/envs/dtailor27/bin/python\r\n"
     ]
    }
   ],
   "source": [
    "!which python"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# D-Tailor testing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Solution class\n",
    "- The sequence and all of its features stored in a single class\n",
    "- Troubleshooting\n",
    "    - ``Features.Structure.StructureMFE``?\n",
    "        - Runs ``Functions.analyze_structure_mfe``\n",
    "        - Which executes ``./3rdParty/unafold/ct-energy``\n",
    "    - ``Features.Structure.StructureDoubleStranded``?\n",
    "        - Runs ``Functions.analyze_structure_ds``\n",
    "        - Which executes ``perl 3rdParty/unafold/ss-count.pl``\n",
    "    - Need to troubleshoot and install properly all the third party programs...currently none working"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'cdsCAI': 0.6136121593930156,\n",
       " 'utrCdsStructureDoubleStranded': 'NA',\n",
       " 'utrCdsStructureDoubleStrandedList': 'NA',\n",
       " 'utrCdsStructureMFE': 0,\n",
       " 'utrNucleotideContentA': 0.45098039215686275,\n",
       " 'utrNucleotideContentAT': 0.6274509803921569,\n",
       " 'utrNucleotideContentC': 0.21568627450980393,\n",
       " 'utrNucleotideContentG': 0.1568627450980392,\n",
       " 'utrNucleotideContentGC': 0.37254901960784315,\n",
       " 'utrNucleotideContentT': 0.17647058823529413}"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Solution import Solution\n",
    "from Features.CAI import CAI\n",
    "from Features.Structure import Structure, StructureMFE, StructureDoubleStranded\n",
    "from Features.NucleotideContent import NucleotideContent\n",
    "from Data import cai_table_ec\n",
    "\n",
    "## Instantiate an object of type 'Solution'\n",
    "solution = Solution(sol_id='b0001', \n",
    "                    sequence='TATAGGCATAGCGCACAGACAGATAAAAATTACAGAGTACACAACATCCATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCACAGGTAACGGTGCGGGCTGA')\n",
    "\n",
    "\n",
    "## Instantiate Feature objects of interest\n",
    "\n",
    "# Feature to calculates the codon adaptation index\n",
    "cai_obj = CAI(solution=solution,label=\"cds\",cai_table=cai_table_ec, args= {'cai_range':(49,115)})\n",
    "\n",
    "# Feature to predicts RNA Structure\n",
    "st1_obj = Structure(solution=solution,label=\"utrCds\",args= { 'structure_range' : (19,78) } )\n",
    "\n",
    "# Two sub-features inheriting from the class Structure\n",
    "st_mfe = StructureMFE(st1_obj)\n",
    "st_ss = StructureDoubleStranded(st1_obj)\n",
    "st1_obj.add_subfeature(st_mfe)\n",
    "st1_obj.add_subfeature(st_ss)\n",
    "\n",
    "# Feature to calculate nucleotide content\n",
    "nuc_obj = NucleotideContent(solution=solution ,label=\"utr\",args= { 'ntcontent_range':(0,50) } )\n",
    "\n",
    "\n",
    "## Add features to solution\n",
    "solution.add_feature(cai_obj)\n",
    "solution.add_feature(st1_obj)\n",
    "solution.add_feature(nuc_obj)\n",
    "\n",
    "\n",
    "## Retrieve feature score\n",
    "solution.scores\n",
    "# {'cdsCAI': 0.6136121593930156, 'utrCdsStructureDoubleStrandedList':[18, 19, 25, 26, 38, 39, 44, 45],\n",
    "# 'utrCdsStructureDoubleStranded': 8, 'utrCdsStructureMFE': -2.5,\n",
    "# 'utrNucleotideContentAT': 0.63, 'utrNucleotideContentG': 0.16,'utrNucleotideContentT' : 0.18,\n",
    "# 'utrNucleotideContentC': 0.22, 'utrNucleotideContentA' : 0.45, 'utrNucleotideContentGC': 0.37}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SequenceAnalyzer class\n",
    "- Lets you take in a bunch of sequences and compute the same properties for all of them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_name,sd_hyb_energy,mfe_structure,cai\n",
      "b0001 , NA , 0 , 0.613612159393\n",
      "b0002 , NA , 0 , 0.34043688741\n",
      "b0003 , NA , 0 , 0.341658034933\n",
      "b0004 , NA , 0 , 0.385891327353\n",
      "b0005 , NA , 0 , 0.377281853234\n",
      "b0006 , NA , 0 , 0.342733396212\n",
      "b0007 , NA , 0 , 0.319183029826\n",
      "b0008 , NA , 0 , 0.604195702312\n",
      "b0009 , NA , 0 , 0.396623675448\n",
      "b0010 , NA , 0 , 0.574062247682\n",
      "b0011 , NA , 0 , 0.286738246339\n",
      "b0013 , NA , 0 , 0.362374253526\n",
      "b0014 , NA , 0 , 0.723381361599\n",
      "b0015 , NA , 0 , 0.525547136369\n",
      "b0016 , NA , 0 , 0.224833425679\n",
      "b0018 , NA , 0 , 0.325397473007\n",
      "b4412 , NA , 0 , 0.388829489281\n",
      "b0019 , NA , 0 , 0.315147538129\n",
      "b0020 , NA , 0 , 0.326664585708\n",
      "b0021 , NA , 0 , 0.259970855677\n",
      "b0022 , NA , 0 , 0.35382053189\n",
      "b0023 , NA , 0 , 0.673378474366\n",
      "b0024 , NA , 0 , 0.223455416035\n",
      "b0025 , NA , 0 , 0.341974173688\n",
      "b0026 , NA , 0 , 0.544170808606\n",
      "b0027 , NA , 0 , 0.375930315257\n",
      "b0028 , NA , 0 , 0.446968612102\n",
      "b0029 , NA , 0 , 0.43576259638\n",
      "b0030 , NA , 0 , 0.33007758527\n",
      "b0031 , NA , 0 , 0.360583601428\n",
      "b0032 , NA , 0 , 0.419594043669\n",
      "b0033 , NA , 0 , 0.565238007841\n",
      "b0034 , NA , 0 , 0.274217763328\n",
      "b0035 , NA , 0 , 0.244827160564\n",
      "b0036 , NA , 0 , 0.403165554531\n",
      "b0037 , NA , 0 , 0.339820638368\n",
      "b0038 , NA , 0 , 0.452476250935\n",
      "b0039 , NA , 0 , 0.476208287595\n",
      "b0040 , NA , 0 , 0.389920755094\n",
      "b0041 , NA , 0 , 0.387332713602\n",
      "b0042 , NA , 0 , 0.467128281643\n",
      "b0043 , NA , 0 , 0.387986561249\n",
      "b0044 , NA , 0 , 0.398836859118\n",
      "b0045 , NA , 0 , 0.298070086981\n",
      "b0046 , NA , 0 , 0.358449938127\n",
      "b0047 , NA , 0 , 0.339327831857\n",
      "b0048 , NA , 0 , 0.400136382299\n",
      "b0049 , NA , 0 , 0.389342815555\n",
      "b0050 , NA , 0 , 0.29573399822\n",
      "b0051 , NA , 0 , 0.373683932955\n",
      "b0052 , NA , 0 , 0.307689655904\n"
     ]
    }
   ],
   "source": [
    "from SequenceAnalyzer import SequenceAnalyzer\n",
    "from Features import CAI,Structure,RNADuplex\n",
    "from Functions import validateCDS\n",
    "from Data import cai_table_ec\n",
    "\n",
    "class TranslationFeaturesEcoliAnalyzer(SequenceAnalyzer):\n",
    "\n",
    "    '''\n",
    "    Class to analyze CAI, SD strength and structure in E. coli\n",
    "    '''\n",
    "\n",
    "    def __init__(self, input_file, input_type):\n",
    "        SequenceAnalyzer.__init__(self,input_file,input_type)\n",
    "\n",
    "    def configureSolution(self, solution):\n",
    "        solution.valid = validateCDS(solution.sequence[49:])\n",
    "        if solution.valid:\n",
    "            #CAI\n",
    "            cai_obj = CAI.CAI(solution=solution,\n",
    "                              label=\"cds\",\n",
    "                              cai_table = cai_table_ec,\n",
    "                              args= { 'cai_range' :(49,len(solution.sequence)) } )\n",
    "\n",
    "            #Look for RBS\n",
    "            dup_obj1 = RNADuplex.RNADuplexRibosome(solution1=solution, \n",
    "                                                   label=\"sd16s\",\n",
    "                                                   args = { 'rnaMolecule1region' : (25,48) })\n",
    "            dup_mfe = RNADuplex.RNADuplexMFE(dup_obj1)\n",
    "            dup_obj1.add_subfeature(dup_mfe)\n",
    "\n",
    "            #MFE [-30,30]\n",
    "            st1_obj = Structure.Structure(solution=solution,\n",
    "                                          label=\"utr\", \n",
    "                                          args= { 'structure_range' : (49-30,49+30) } )\n",
    "            st_mfe = Structure.StructureMFE(st1_obj)\n",
    "            st1_obj.add_subfeature(st_mfe)\n",
    "\n",
    "            solution.add_feature(cai_obj)\n",
    "            solution.add_feature(dup_obj1)\n",
    "            solution.add_feature(st1_obj)\n",
    "    \n",
    "    def outputStart(self):\n",
    "        print \"gene_name,sd_hyb_energy,mfe_structure,cai\"\n",
    "\n",
    "    def output(self, solution):\n",
    "        if solution.valid:\n",
    "            print solution.solid,',',\n",
    "            print solution.scores['sd16sRNADuplexMFE'],',',\n",
    "            print solution.scores['utrStructureMFE'],',',\n",
    "            print solution.scores['cdsCAI']\n",
    "\n",
    "seqAnalyzerTest = TranslationFeaturesEcoliAnalyzer(\"testFiles/genomes/partial_ecoli_genome.csv\",\"CSV\")\n",
    "seqAnalyzerTest.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SequenceDesigner class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def configureSolution(self, solution):\n",
    "    '''\n",
    "    Solution configuration\n",
    "    '''\n",
    "    if solution.sequence == None:\n",
    "        return 0\n",
    "\n",
    "    ## Designer specific\n",
    "    solution.mutable_region=range(0,len(solution.sequence)) # whole region\n",
    "    solution.cds_region = (49,len(solution.sequence))\n",
    "    solution.keep_aa = True\n",
    "\n",
    "    ## Populate solution with desired features\n",
    "\n",
    "    # CAI\n",
    "    cai_obj = CAI.CAI(solution = solution,\n",
    "                      label=\"cds\",\n",
    "                      cai_table = cai_table_ec,\n",
    "                      args = {'cai_range': (49,len(solution.sequence)),\n",
    "                              'mutable_region': range(49,len(solution.sequence)) } )\n",
    "\n",
    "    # Search SD\n",
    "    dup_obj1 = RNADuplex.RNADuplexRibosome(solution1=solution, label=\"sd16s\",\n",
    "                                            args = { 'rnaMolecule1region' : (25,48),\n",
    "                                            'mutable_region' : range(25,48) })\n",
    "    dup_mfe = RNADuplex.RNADuplexMFE(dup_obj1)\n",
    "    dup_obj1.add_subfeature(dup_mfe)\n",
    "\n",
    "    # MFE [-30,30]\n",
    "    st1_obj = Structure.Structure(solution=solution,label=\"utr\",\n",
    "                                args = {'structure_range' : (49-30,49+30), \n",
    "                                        'mutable_region' : range(49-30,49+30)} )\n",
    "    st_mfe = Structure.StructureMFE(st1_obj)\n",
    "    st1_obj.add_subfeature(st_mfe)\n",
    "\n",
    "    solution.add_feature(cai_obj)\n",
    "    solution.add_feature(dup_obj1)\n",
    "    solution.add_feature(st1_obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1.1.1',\n",
       " '1.1.3',\n",
       " '1.1.2',\n",
       " '1.1.5',\n",
       " '1.1.4',\n",
       " '1.3.1',\n",
       " '1.3.3',\n",
       " '1.3.2',\n",
       " '1.3.5',\n",
       " '1.3.4',\n",
       " '1.2.1',\n",
       " '1.2.3',\n",
       " '1.2.2',\n",
       " '1.2.5',\n",
       " '1.2.4',\n",
       " '1.5.1',\n",
       " '1.5.3',\n",
       " '1.5.2',\n",
       " '1.5.5',\n",
       " '1.5.4',\n",
       " '1.4.1',\n",
       " '1.4.3',\n",
       " '1.4.2',\n",
       " '1.4.5',\n",
       " '1.4.4',\n",
       " '3.1.1',\n",
       " '3.1.3',\n",
       " '3.1.2',\n",
       " '3.1.5',\n",
       " '3.1.4',\n",
       " '3.3.1',\n",
       " '3.3.3',\n",
       " '3.3.2',\n",
       " '3.3.5',\n",
       " '3.3.4',\n",
       " '3.2.1',\n",
       " '3.2.3',\n",
       " '3.2.2',\n",
       " '3.2.5',\n",
       " '3.2.4',\n",
       " '3.5.1',\n",
       " '3.5.3',\n",
       " '3.5.2',\n",
       " '3.5.5',\n",
       " '3.5.4',\n",
       " '3.4.1',\n",
       " '3.4.3',\n",
       " '3.4.2',\n",
       " '3.4.5',\n",
       " '3.4.4',\n",
       " '2.1.1',\n",
       " '2.1.3',\n",
       " '2.1.2',\n",
       " '2.1.5',\n",
       " '2.1.4',\n",
       " '2.3.1',\n",
       " '2.3.3',\n",
       " '2.3.2',\n",
       " '2.3.5',\n",
       " '2.3.4',\n",
       " '2.2.1',\n",
       " '2.2.3',\n",
       " '2.2.2',\n",
       " '2.2.5',\n",
       " '2.2.4',\n",
       " '2.5.1',\n",
       " '2.5.3',\n",
       " '2.5.2',\n",
       " '2.5.5',\n",
       " '2.5.4',\n",
       " '2.4.1',\n",
       " '2.4.3',\n",
       " '2.4.2',\n",
       " '2.4.5',\n",
       " '2.4.4',\n",
       " '5.1.1',\n",
       " '5.1.3',\n",
       " '5.1.2',\n",
       " '5.1.5',\n",
       " '5.1.4',\n",
       " '5.3.1',\n",
       " '5.3.3',\n",
       " '5.3.2',\n",
       " '5.3.5',\n",
       " '5.3.4',\n",
       " '5.2.1',\n",
       " '5.2.3',\n",
       " '5.2.2',\n",
       " '5.2.5',\n",
       " '5.2.4',\n",
       " '5.5.1',\n",
       " '5.5.3',\n",
       " '5.5.2',\n",
       " '5.5.5',\n",
       " '5.5.4',\n",
       " '5.4.1',\n",
       " '5.4.3',\n",
       " '5.4.2',\n",
       " '5.4.5',\n",
       " '5.4.4',\n",
       " '4.1.1',\n",
       " '4.1.3',\n",
       " '4.1.2',\n",
       " '4.1.5',\n",
       " '4.1.4',\n",
       " '4.3.1',\n",
       " '4.3.3',\n",
       " '4.3.2',\n",
       " '4.3.5',\n",
       " '4.3.4',\n",
       " '4.2.1',\n",
       " '4.2.3',\n",
       " '4.2.2',\n",
       " '4.2.5',\n",
       " '4.2.4',\n",
       " '4.5.1',\n",
       " '4.5.3',\n",
       " '4.5.2',\n",
       " '4.5.5',\n",
       " '4.5.4',\n",
       " '4.4.1',\n",
       " '4.4.3',\n",
       " '4.4.2',\n",
       " '4.4.5',\n",
       " '4.4.4']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from DesignOfExperiments.Design import FullFactorial\n",
    "\n",
    "#Design Methodology and thresholds\n",
    "design_param = { \"sd16sRNADuplexMFE\": {'type' : 'REAL' ,\n",
    "                                       'thresholds' : {'1': (-12.7,-7.3), '2': (-7.3,-5.8),\n",
    "                                                       '3': (-5.8,-5.2), '4': (-5.2,-3.3), '5': (-3.3, 2.0) } },\n",
    "                \"utrStructureMFE\": {'type' : 'REAL' ,\n",
    "                                    'thresholds' : {'1': (-29.2,-12.2), '2': (-12.2,-9.95),\n",
    "                                                    '3': (-9.95,-8.4), '4': (-8.4,-6.73), '5': (-6.73,0.65) } },\n",
    "                \"cdsCAI\" : {'type' : 'REAL' ,\n",
    "                            'thresholds' : {'1': (0.13,0.29), '2': (0.29,0.33),\n",
    "                                            '3': (0.33,0.37), '4': (0.37,0.42), '5': (0.42,0.86) } } }\n",
    "design = FullFactorial([\"sd16sRNADuplexMFE\",\"utrStructureMFE\",\"cdsCAI\"],design_param)\n",
    "design.listDesigns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "looking for combination:  5\n",
      "Solution found... inserting into DB...\n",
      "\n",
      "###########################\n",
      "# Optimized solution:\n",
      "# ID:  330648123324667243168199564780158269150\n",
      "# Sequence:  ttattaccggacaataatatttcaattcattaaagaggagaaaggtaccatggcgctgtggatgcgcctcctgcctctgctggcgctgctggccctctggggacctgacccggccgcagcttttgtgaaccaacacctgtgcggcagccacctggtggaagctctctacctagtgtgcggggaacgtggcttcttctacacacctaagacccgccgggaggcagaggacctgcaggtgggtcaggtagaactgggcggtggccctggtgcaggtagcctgcagcccttggctctggaggggtccctgcagaagcgtggcattgtagaacaatgctgtaccagcatctgctccctctaccagctggagaactactgcaactag\n",
      "# Scores:  ['cdsCAI: 0.42264774749']\n",
      "# Levels:  ['cdsCAILevel: 5']\n",
      "# Number of generated solutions:  17\n",
      "# Distance to seed:  16\n",
      "###########################\n",
      "\n",
      "Program finished, all combinations were found...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(17, 16, 3.141724799297725)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from RunningExamples.Designer.TranslationFeaturesEcoliDesigner import TranslationFeaturesEcoliDesigner\n",
    "from DesignOfExperiments.Design import FullFactorial, Optimization\n",
    "\n",
    "#Seed sequence from which mutants will be derived\n",
    "seed='ttattaccggacaataatatttcaattcattaaagaggagaaaggtaccatggccctgtggatgcgcctcctgcccctgctggcgctgctggccctctggggacctgacccagccgcagcctttgtgaaccaacacctgtgcggctcacacctggtggaagctctctacctagtgtgcggggaacgaggcttcttctacacacccaagacccgccgggaggcagaggacctgcaggtggggcaggtggagctgggcgggggccctggtgcaggcagcctgcagcccttggccctggaggggtccctgcagaagcgtggcattgtggaacaatgctgtaccagcatctgctccctctaccagctggagaactactgcaactag'\n",
    "\n",
    "#Design Methodology and thresholds\n",
    "design_param = { \n",
    " \"cdsCAI\" : { 'type' : 'REAL' ,\n",
    " 'thresholds' : { '1': (0.13,0.29), '2': (0.29,0.33),\n",
    " '3': (0.33,0.37), '4': (0.37,0.42),\n",
    "'5': (0.42,0.86) } }\n",
    " }\n",
    "\n",
    "design = Optimization([\"cdsCAI\"],design_param, '5')\n",
    "\n",
    "# root_dir = \"/Users/nmih/Projects/\"\n",
    "root_dir = \"/vagrant/\"\n",
    "output_file = root_dir + 'codon_optimization/d-tailor/testFiles/outputFiles/tfec_1'\n",
    "\n",
    "tirap_designer = TranslationFeaturesEcoliDesigner(\"tfec\", \n",
    "                                                  seed, \n",
    "                                                  design,\n",
    "                                                  output_file, \n",
    "                                                  createDB=True)\n",
    "tirap_designer.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Own code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating CAI for Pk\n",
    "\n",
    "- Just use the simple Biopython calculator - **but needed to modify code (Bio.SeqUtils.CodonUsage) to get it to skip the unknown bases (N)** REMEMBER THIS!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqUtils.CodonUsage import CodonAdaptationIndex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "illegal codon ACN in gene: lcl|NW_018150436.1_cds_XP_020546087.1_933\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon NNN in gene: lcl|NW_018150437.1_cds_XP_020545573.1_1020\n",
      "illegal codon CGN in gene: lcl|NW_018150439.1_cds_XP_020545092.1_1667\n",
      "illegal codon TA in gene: lcl|NW_018150441.1_cds_XP_020544721.1_2022\n",
      "illegal codon GAN in gene: lcl|NW_018150443.1_cds_XP_020544345.1_2469\n",
      "illegal codon GAN in gene: lcl|NW_018150449.1_cds_XP_020543512.1_3267\n",
      "illegal codon AG in gene: lcl|NW_018150449.1_cds_XP_020543561.1_3371\n",
      "illegal codon GNC in gene: lcl|NW_018150463.1_cds_XP_020542560.1_4293\n",
      "illegal codon AA in gene: lcl|NW_018150484.1_cds_XP_020541991.1_4888\n",
      "illegal codon GA in gene: lcl|NW_018150525.1_cds_XP_020541789.1_5077\n",
      "illegal codon A in gene: lcl|NW_018150531.1_cds_XP_020541769.1_5099\n",
      "illegal codon AG in gene: lcl|NW_018150535.1_cds_XP_020541760.1_5107\n",
      "illegal codon AT in gene: lcl|NW_018150536.1_cds_XP_020541754.1_5111\n",
      "illegal codon A in gene: lcl|NW_018150539.1_cds_XP_020541749.1_5118\n",
      "illegal codon C in gene: lcl|NW_018150550.1_cds_XP_020541720.1_5148\n",
      "illegal codon G in gene: lcl|NW_018150603.1_cds_XP_020541604.1_5265\n",
      "illegal codon TA in gene: lcl|NW_018150630.1_cds_XP_020541556.1_5312\n",
      "illegal codon A in gene: lcl|NW_018150642.1_cds_XP_020541541.1_5328\n",
      "illegal codon TT in gene: lcl|NW_018150666.1_cds_XP_020541510.1_5359\n",
      "illegal codon A in gene: lcl|NW_018150689.1_cds_XP_020541486.1_5382\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'aaa': 1.0,\n",
       " 'aac': 0.6835958956781579,\n",
       " 'aag': 0.8501819970486965,\n",
       " 'aat': 1.0,\n",
       " 'aca': 0.9698236035279295,\n",
       " 'acc': 0.6997060058798824,\n",
       " 'acg': 0.3369172616547669,\n",
       " 'act': 1.0,\n",
       " 'aga': 1.0,\n",
       " 'agc': 0.33871546465733765,\n",
       " 'agg': 0.4327671815167005,\n",
       " 'agt': 0.6517581628991748,\n",
       " 'ata': 0.44500644063390016,\n",
       " 'atc': 0.560287878608823,\n",
       " 'atg': 1.0,\n",
       " 'att': 1.0,\n",
       " 'caa': 1.0,\n",
       " 'cac': 0.4904431664411366,\n",
       " 'cag': 0.5018068436200831,\n",
       " 'cat': 1.0,\n",
       " 'cca': 1.0,\n",
       " 'ccc': 0.27423029492297724,\n",
       " 'ccg': 0.26806429362707185,\n",
       " 'cct': 0.6827540079008424,\n",
       " 'cga': 0.13356785576870225,\n",
       " 'cgc': 0.059216914930650975,\n",
       " 'cgg': 0.11079352864654904,\n",
       " 'cgt': 0.2555012000952747,\n",
       " 'cta': 0.45494302516671425,\n",
       " 'ctc': 0.2547365847695881,\n",
       " 'ctg': 0.26136440804574285,\n",
       " 'ctt': 0.4427739070203317,\n",
       " 'gaa': 1.0,\n",
       " 'gac': 0.4646572115247604,\n",
       " 'gag': 0.5341032773007619,\n",
       " 'gat': 1.0,\n",
       " 'gca': 1.0,\n",
       " 'gcc': 0.4917465037926556,\n",
       " 'gcg': 0.2468331796194169,\n",
       " 'gct': 0.8223192606674321,\n",
       " 'gga': 0.5128889736499158,\n",
       " 'ggc': 0.3511013635930199,\n",
       " 'ggg': 0.27271860398588726,\n",
       " 'ggt': 1.0,\n",
       " 'gta': 0.2980785532636338,\n",
       " 'gtc': 0.45507205425261377,\n",
       " 'gtg': 0.3633088443063012,\n",
       " 'gtt': 1.0,\n",
       " 'taa': 1.0,\n",
       " 'tac': 0.6870831268818843,\n",
       " 'tag': 0.6097455799913757,\n",
       " 'tat': 1.0,\n",
       " 'tca': 0.8974345174022247,\n",
       " 'tcc': 0.6188733405095085,\n",
       " 'tcg': 0.4076785073555795,\n",
       " 'tct': 1.0,\n",
       " 'tga': 0.6722725312634756,\n",
       " 'tgc': 0.41045642382875985,\n",
       " 'tgg': 1.0,\n",
       " 'tgt': 1.0,\n",
       " 'tta': 0.8421410042238792,\n",
       " 'ttc': 0.649768556417164,\n",
       " 'ttg': 1.0,\n",
       " 'ttt': 1.0}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pk_cds = '/vagrant/genomes/Pk/GCF_001983325.1_ASM198332v1_cds_from_genomic.fna'\n",
    "\n",
    "biop_cai = CodonAdaptationIndex()\n",
    "biop_cai.generate_index(fasta_file=pk_cds)\n",
    "\n",
    "PK_CAI_TABLE = {k.lower():v for k,v in biop_cai.index.items()}\n",
    "PK_CAI_TABLE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating CAI quartiles for Pk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get all coding sequences\n",
    "\n",
    "sequences = []\n",
    "with open(pk_cds) as f:\n",
    "    for s in SeqIO.parse(f, 'fasta'):\n",
    "        sequences.append(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate CAIs\n",
    "\n",
    "from Solution import Solution\n",
    "from Features.CAI import CAI\n",
    "\n",
    "all_pk_cais = {}\n",
    "\n",
    "for s in sequences:\n",
    "    solution = Solution(sol_id=s.id, \n",
    "                        sequence=str(s.seq))\n",
    "    cai_obj = CAI(solution=solution,label=\"cds\",cai_table=PK_CAI_TABLE, args= {'cai_range':(0,len(s))})\n",
    "    solution.add_feature(cai_obj)\n",
    "    all_pk_cais[solution.solid] = solution.scores['cdsCAI']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1': (0, 0.4515777307310756),\n",
       " '2': (0.4515777307310756, 0.7025300201378566),\n",
       " '3': (0.7025300201378566, 0.7323131353898734),\n",
       " '4': (0.7323131353898734, 0.7590354619135496),\n",
       " '5': (0.7590354619135496, 1)}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get quintiles for thresholds\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "pk_cai_quintiles = []\n",
    "for x in [0, 25, 50, 75, 100]:\n",
    "    pk_cai_quintiles.append(np.percentile(list(all_pk_cais.values()), x))\n",
    "\n",
    "pk_cai_thresholds = {}\n",
    "for i, x in enumerate(pk_cai_quintiles):\n",
    "    if i == 0:\n",
    "        pk_cai_thresholds[str(i+1)] = (0, x)\n",
    "    elif i == len(pk_cai_quintiles) - 1:\n",
    "        pk_cai_thresholds[str(i+1)] = (pk_cai_quintiles[i-1], 1)\n",
    "    else:\n",
    "        pk_cai_thresholds[str(i+1)] = (pk_cai_quintiles[i-1], x)\n",
    "\n",
    "pk_cai_thresholds"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running CAI maximization for Pk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Functions\n",
    "import sys\n",
    "from SequenceDesigner import SequenceDesigner\n",
    "from Features.Structure import Structure,StructureMFE\n",
    "from Features import CAI,RNADuplex\n",
    "from DesignOfExperiments.Design import RandomSampling,Optimization,FullFactorial\n",
    "\n",
    "class TranslationFeaturesPkDesigner(SequenceDesigner):\n",
    "    \n",
    "    def __init__(self, name, seed, design, dbfile, createDB=True):\n",
    "        SequenceDesigner.__init__(self, name, seed, design, dbfile, createDB)\n",
    "        \n",
    "    def configureSolution(self, solution):\n",
    "        \"\"\"\n",
    "        Solution configuration\n",
    "        \n",
    "        \"\"\"\n",
    "                \n",
    "        if solution.sequence == None:\n",
    "            return 0\n",
    "        \n",
    "        # TODO: need to replace these numbers and auto calculate them\n",
    "        mutable_start = 0\n",
    "        mutable_end = len(solution.sequence)\n",
    "        cds_start = 49\n",
    "        cds_end = len(solution.sequence)\n",
    "        \n",
    "        # Populate solution with desired features\n",
    "        solution.mutable_region = range(mutable_start, mutable_end)\n",
    "        solution.cds_region = (cds_start, cds_end)\n",
    "        \n",
    "        # This should always be true unless you want to make mutant proteins\n",
    "        solution.keep_aa = True\n",
    "        \n",
    "        cai_obj = CAI.CAI(solution=solution,\n",
    "                          label=\"cds\",\n",
    "                          cai_table=PK_CAI_TABLE, \n",
    "                          args={'cai_range' : solution.cds_region, \n",
    "                                'mutable_region' : solution.mutable_region } )\n",
    "            \n",
    "        solution.add_feature(cai_obj)                \n",
    "    \n",
    "    def validateSolution(self, solution):\n",
    "        '''\n",
    "        Solution validation tests\n",
    "        \n",
    "        Additionally, the user may also want to implement the method \n",
    "        validateSolution, which is called every time a new\n",
    "        sequence is generated. This validation step is fundamental\n",
    "        to avoid undesired properties in new sequence variants \n",
    "        (e.g., a spurious restriction site). Only validated sequences will\n",
    "        be stored in the database. \n",
    "\n",
    "        '''\n",
    "        if solution.sequence == None or ('?' in solution.levels.values()):\n",
    "            sys.stderr.write(\"SolutionValidator: Level unknown - \"+str(solution.levels)+\"\\n\")                        \n",
    "            solution.valid = False\n",
    "            return 0\n",
    "        \n",
    "        # Check if solution is valid\n",
    "        valid = True\n",
    "      \n",
    "        designed_region = solution.sequence\n",
    "                \n",
    "#         # No internal Promoters\n",
    "#         (score, _, _) = Functions.look_for_promoters(designed_region)\n",
    "#         if score >= 15.3990166: #0.95 percentile for Promoter PWM scores\n",
    "#             valid = False\n",
    "#             sys.stderr.write(\"SolutionValidator: High Promoter score: \"+str(score)+\"\\n\")                    \n",
    "        \n",
    "#         # No internal Terminator\n",
    "#         score = Functions.look_for_terminators(designed_region)\n",
    "#         if score >= 90: #90% confidence from transtermHP\n",
    "#             valid = False\n",
    "#             sys.stderr.write(\"SolutionValidator: High Terminator score\\n\")    \n",
    "            \n",
    "#         # No restriction enzymes\n",
    "#         if 'ggtctc' in designed_region or 'gagacc' in designed_region:\n",
    "#             sys.stderr.write(\"SolutionValidator: Restriction enzyme found\\n\")\n",
    "#             valid = False        \n",
    "        \n",
    "        solution.valid = valid\n",
    "        \n",
    "        return valid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Starting CAI: ', 0.6402064678148808)\n",
      "looking for combination:  5\n",
      "No solution could be found...\n",
      "looking for combination:  5\n",
      "No solution could be found...\n",
      "looking for combination:  5\n",
      "No solution could be found...\n",
      "looking for combination:  5\n",
      "No solution could be found...\n",
      "looking for combination:  5\n",
      "No solution could be found...\n",
      "looking for combination:  5\n",
      "No solution could be found...\n",
      "looking for combination:  5\n",
      "No solution could be found...\n",
      "looking for combination:  5\n",
      "No solution could be found...\n",
      "looking for combination:  5\n",
      "No solution could be found...\n",
      "looking for combination:  5\n",
      "No solution could be found...\n",
      "looking for combination:  5\n",
      "No solution could be found...\n",
      "looking for combination:  5\n",
      "No solution could be found...\n",
      "looking for combination:  5\n",
      "No solution could be found...\n",
      "looking for combination:  5\n",
      "No solution could be found...\n",
      "looking for combination:  5\n",
      "No solution could be found...\n",
      "looking for combination:  5\n",
      "No solution could be found...\n",
      "time elapsed: 60.99 (s) \t solutions generated: 1585 \t rate (last min.): 25.97 sol/s  \t rate (overall): 25.99 sol/s\n",
      "looking for combination:  5\n",
      "No solution could be found...\n",
      "looking for combination:  5\n",
      "No solution could be found...\n",
      "looking for combination:  5\n",
      "No solution could be found...\n",
      "looking for combination:  5\n",
      "Solution found... inserting into DB...\n",
      "\n",
      "###########################\n",
      "# Optimized solution:\n",
      "# ID:  118233300767199293493876444068448429182\n",
      "# Sequence:  atgagtccaagcaccgctccggcagttaaaccagacgctggtgactcatttggcataataggaaaactaatcaccatcttaaaagcaatggcctgctcaaagagatcccatgagattgttaagatggagcaaaaacaagaacagacgattgaatacgaacacctgattggaaagtctatagtaaagcagagatccatctccaattgattgtaatcagagttagctcaacactattcatcatacctgtcaggattacagaaatactagcaattatacttccaaagtttcgaatcatttaggatcttttttggcaactaagctggtcaattcttaaactggtttgaagacttcgacagcgtgtttgtatttaacccgtagacctgaatctttgcattagagaactatgcatggttcctaaatggcatcttaaatgcggcatactattgtgttgatcgacatgcattataaacgtccaactaaccttccatcatttatgaggttaacgtcacgtgtcagagatactcattaacatatgctgaattgttagaatagatttgtcaggtcgctcagatcttgcagtatagtttgggcgtttgtgagggagattcaatagaggtttatttgccaatgatttcgtaaacattgattgccttaatggcaatctccaggattggaattgtttactctgttgtgtttgttggtttttctagcaactgtctgcgttatcgaattgaccacaccaattccaggttagtctttgccaaggacgagtccaaacgtcggtgtaaaatccttgagaccaaaagaaagttctatgacgttgtaaaggaaacatcaagggtctgaaatgtacttgtatattagagaacaaatcattttaaggtgttcttggtctcagggcgtgatctggattggaacgaccagattgaaaaatataaaggatactgcacctgcaaatctttggatagcgtccactcgcttttcttattatactctacaagcattaagagttttgccaaaagcctttaatactgcacttcaagttatttgttaggttctcttctgaccatgtgttactcatagtactcccacaggcaagacgtatttttcacgtcccgtgatgtgggcaggattacttgacactccaactttgtttacgggtccatattatacggttgttcaacacttgtatagtaaggcattgccgccaatccaaactactttagttattgggatatcttagactaatataaagtcacgaaattttatgtggcttccaccgagattagattgttgaagaggtcttggcactttgacgtagagaacaactcattaaaatcattaaggtgtttaggtcctgtaggaatccccatctcagccaaagtgtggtactggtactgcatccaaatccgttaaaatgagattcttgtagtagatcttaattggcaaacaaagtttgggtgctatatgttaactcaggtagcaaggtgaattgcaccgttgaagtcagggtgcgcatgctttcccttctttggtttagaaccaattatattggatgagaactcgggaaaaaaaattgcagagacatatgcttaaagtctattggcgatgaaatccagatggtcctcctttgagagaacgatatggaaaaattatgatcggttcttagatacgtacatgaagtccatcaaagggtattacttctcaggagatggtgccaccagggattaagatggatatatgtggttcttgggcagggtcgacgacttcttaaatgtgtccagacactggatgtccaaggcaaaaattgtctctgcaatcttagactacagcattgtagcaatctgttagatggtgggtttctatgatgacctgacttgataggctgtggcggccatcttagtgttgaagaacttctcatccaggagcattaccaccaaatccgagttataggactttgaaaaacatttgattttggagatgagaaaagacatcggatccattgcagcccccaaattgatagtactggttgacgatttatccaagactggttcaagttaaattgtgaggtgaatactgtgtgaaatactagctggaaaaagcaatcaattaggcgacttcagtaccatatcaaattcgggcattgtctgttactttgtagactcaataatcatataa\n",
      "# Scores:  ['cdsCAI: 0.759558875574']\n",
      "# Levels:  ['cdsCAILevel: 5']\n",
      "# Number of generated solutions:  1960\n",
      "# Distance to seed:  315\n",
      "###########################\n",
      "\n",
      "Program finished, all combinations were found...\n",
      "CPU times: user 1min 2s, sys: 4.88 s, total: 1min 7s\n",
      "Wall time: 1min 15s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "from DesignOfExperiments.Design import FullFactorial, Optimization\n",
    "\n",
    "# Seed sequence from which mutants will be derived\n",
    "seed = 'ATGAGTCCAAGCACCGCTCCGGCAGTTAAACCAGACGCTGGTGACTCATCTAGCATAATAGGAGAACTAAAGTCCAAGCTAAAAGCAGTGGCCGGCTCAGAGAGATCCGATGAGGTTGCTGAGATGGAGCAAGAACAAGAACAGGCGACTGAATACGAACACCTGACTAGAGTCCCTATAGTAAAGCAGAGACCCATCTCCGATAGATTGCAACCAGAGTTAGCTCAACACTATTCACCACACCTGTCAGGATTACAGGAATACAAGCAATTATACAAGCAAAGTATCGAATCACCTGGGAAGTTTTTTGGCGACAAAGCTCGTCAATTCTTAAACTGGTTTAAAGACTTCGACAGCGTGTTTGTACCTAACCCGGAGACCGGAAAGCCTTCATTAGAGAACAATGCATGGTTCCTAAATGGCCAGTTAAATGCGGCATACAATTGTGTTGATAGACATGCACTACAAACGCCCAACAAACCTGCCATCATTTATGAGGCTGACGAGCCGGGTCAGGGATACACACTAACATATGCTGAATTGTTAGAACAGGTTTGTAAGGTCGCTCAGGTCTTGCAGTATAGTATGGGCGTTCGTAAGGGAGATACAGTAGCGGTTTATATGCCAATGATTCCGCAAACACTGATTACCTTAATGGCAATCACCAGGATTGGAGCTATTCACTCTGTTGTGTTTGCTGGTTTTTCTTGCAACAGTCTGCGTGATAGAATTAACGACGCCGATTCCAGGGTAGTCATTACCACGGACGAGTCCAAACGTGGGGGTAAAATCGTTGAGACCAAAAGAATCGTCGATGACGCTATAAAGGAAACATCAGGGGTCAGAAATGTACTTGTATATAAGAGAACAAATAATCCTAAGGTGCAGCTGGTCCCAGGGCGTGATCTGGATTGGGACGACGAGATTAAAAAATATAAAGGATACTGCCCCTGCGAACCTGTGGATAGCGAGCACCCGCTTTTCTTACTATACACTTCAGGCTCTACGGGTACTCCCAAAGGCGTTCAACACAGCACTGCAGGTTATTTGTTAGGTGCTCTTCTGACCATGCGTTACACATTCGACACCCACAGGGAAGACGTATTTTTCACGGCCGGTGATGTGGGCTGGATTACTGGACACACCTACGTTGTTTACGGGCCCCTATTATACGGTTGTACAACACTTGTATTCGAAGGCACTCCCGCCTATCCAAACTACTCTCGTTATTGGGATATCGTAGACAAATATAAAGTCACGCAATTTTATGTGGCTCCCACCGCGCTTAGATTGTTGAAGAGGGCTGGGGACTCTTACGTAGAGAACCACTCACTAAAATCACTAAGGTGTTTAGGTACTGTAGGAGAGCCCATCGCAGCCGAAGTGTGGGACTGGTACAGCGAGCAAATCGGTAAAAATGAGATTCCTATAGTAGATACTTATTGGCAAACAGAGTCTGGGAGCCATATGTTAACTCCGCTAGCAGGGGGAGTTACACCGATGAAGCCAGGGAGCGCATGCTTTCCCTTCTTTGGTATAGAACCAGCTATACTGGATGCGAACACGGGAAAAGAAATTACAGCGTCACATGCTGAAGGTGTACTGGCGGTGAAATCCGGATGGCCCTCCTTTGCGAGAACGATATGGAAAAATCATGATAGGTTCTTAGATACGTACCTGAAGCCCTTCAAAGGGTATTACTTCGCAGGAGATGGTGCCGCCAGGGATAAAGATGGATATATGTGGATCTTGGGCAGGGTCGACGACGTCGTAAATGTGTCCGGACACAGGCTGTCCACGGCAGAAATTGAGTCTGCAATCCTAGACGACAGCATTGTAGCAGAGTGTGCGGTGGTGGGTTTCAATGATGACCTGACTGGACAGGCTGTGGCGGCCTTCGTAGTGTTGAAGAACAAGTCATCCTGGAGCACTGCCTCCGAATCCGAGTTACAGGACATTAAAAAACATTTGATTTTGGCGGTGAGAAAAGACATCGGACCCTTTGCAGCCCCCAAATTGATAGTACTGGTTGACGATCTACCCAAGACTCGTTCAGGTAAAATTATGAGGAGAATACTGCGTAAAATACTAGCTGGAGAAAGCGATCAATTAGGCGACGTCAGTACCTTATCAAATCCGGGCATTGTCCGTCACCTTATAGACTCAGTAAAGCTATAA'\n",
    "\n",
    "\n",
    "# Design Methodology and thresholds\n",
    "# A set of properties and their respective levels need to be inputted to instantiate a\n",
    "# sub-class of Design. This is given in the form of a dictionary, where for each property it\n",
    "# is necessary to define a type (REAL, INTEGER or TEXT) and a list of levels containing\n",
    "# the respective lower and upper bounds\n",
    "design_param = {\n",
    "    \"cdsCAI\": {\n",
    "        'type'      : 'REAL',\n",
    "        'thresholds': pk_cai_thresholds\n",
    "    }\n",
    "}\n",
    "\n",
    "# Specify the parameters you want to optimize to\n",
    "design = Optimization(featuresList=[\"cdsCAI\"], featuresObj=design_param, target='5')\n",
    "\n",
    "# This is where the SQLite database will be saved\n",
    "root_dir = \"/vagrant/\"\n",
    "output_file = root_dir + 'codon_optimization/d-tailor/testFiles/outputFiles/tfpk_1'\n",
    "\n",
    "# Get ready to run everything\n",
    "tirap_designer = TranslationFeaturesPkDesigner(\"tfpk\",\n",
    "                                               seed,\n",
    "                                               design,\n",
    "                                               output_file,\n",
    "                                               createDB=True)\n",
    "\n",
    "# Print original CAI\n",
    "solution = Solution(sol_id='tmp', \n",
    "                        sequence=seed)\n",
    "cai_obj = CAI.CAI(solution=solution, label=\"cds\", cai_table=PK_CAI_TABLE, args= {'cai_range':(0,len(seed))})\n",
    "solution.add_feature(cai_obj)\n",
    "print('Starting CAI: ', solution.scores['cdsCAI'])\n",
    "\n",
    "# Run it\n",
    "result = tirap_designer.run()\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "orig = Seq(\"ATGAGTCCAAGCACCGCTCCGGCAGTTAAACCAGACGCTGGTGACTCATCTAGCATAATAGGAGAACTAAAGTCCAAGCTAAAAGCAGTGGCCGGCTCAGAGAGATCCGATGAGGTTGCTGAGATGGAGCAAGAACAAGAACAGGCGACTGAATACGAACACCTGACTAGAGTCCCTATAGTAAAGCAGAGACCCATCTCCGATAGATTGCAACCAGAGTTAGCTCAACACTATTCACCACACCTGTCAGGATTACAGGAATACAAGCAATTATACAAGCAAAGTATCGAATCACCTGGGAAGTTTTTTGGCGACAAAGCTCGTCAATTCTTAAACTGGTTTAAAGACTTCGACAGCGTGTTTGTACCTAACCCGGAGACCGGAAAGCCTTCATTAGAGAACAATGCATGGTTCCTAAATGGCCAGTTAAATGCGGCATACAATTGTGTTGATAGACATGCACTACAAACGCCCAACAAACCTGCCATCATTTATGAGGCTGACGAGCCGGGTCAGGGATACACACTAACATATGCTGAATTGTTAGAACAGGTTTGTAAGGTCGCTCAGGTCTTGCAGTATAGTATGGGCGTTCGTAAGGGAGATACAGTAGCGGTTTATATGCCAATGATTCCGCAAACACTGATTACCTTAATGGCAATCACCAGGATTGGAGCTATTCACTCTGTTGTGTTTGCTGGTTTTTCTTGCAACAGTCTGCGTGATAGAATTAACGACGCCGATTCCAGGGTAGTCATTACCACGGACGAGTCCAAACGTGGGGGTAAAATCGTTGAGACCAAAAGAATCGTCGATGACGCTATAAAGGAAACATCAGGGGTCAGAAATGTACTTGTATATAAGAGAACAAATAATCCTAAGGTGCAGCTGGTCCCAGGGCGTGATCTGGATTGGGACGACGAGATTAAAAAATATAAAGGATACTGCCCCTGCGAACCTGTGGATAGCGAGCACCCGCTTTTCTTACTATACACTTCAGGCTCTACGGGTACTCCCAAAGGCGTTCAACACAGCACTGCAGGTTATTTGTTAGGTGCTCTTCTGACCATGCGTTACACATTCGACACCCACAGGGAAGACGTATTTTTCACGGCCGGTGATGTGGGCTGGATTACTGGACACACCTACGTTGTTTACGGGCCCCTATTATACGGTTGTACAACACTTGTATTCGAAGGCACTCCCGCCTATCCAAACTACTCTCGTTATTGGGATATCGTAGACAAATATAAAGTCACGCAATTTTATGTGGCTCCCACCGCGCTTAGATTGTTGAAGAGGGCTGGGGACTCTTACGTAGAGAACCACTCACTAAAATCACTAAGGTGTTTAGGTACTGTAGGAGAGCCCATCGCAGCCGAAGTGTGGGACTGGTACAGCGAGCAAATCGGTAAAAATGAGATTCCTATAGTAGATACTTATTGGCAAACAGAGTCTGGGAGCCATATGTTAACTCCGCTAGCAGGGGGAGTTACACCGATGAAGCCAGGGAGCGCATGCTTTCCCTTCTTTGGTATAGAACCAGCTATACTGGATGCGAACACGGGAAAAGAAATTACAGCGTCACATGCTGAAGGTGTACTGGCGGTGAAATCCGGATGGCCCTCCTTTGCGAGAACGATATGGAAAAATCATGATAGGTTCTTAGATACGTACCTGAAGCCCTTCAAAGGGTATTACTTCGCAGGAGATGGTGCCGCCAGGGATAAAGATGGATATATGTGGATCTTGGGCAGGGTCGACGACGTCGTAAATGTGTCCGGACACAGGCTGTCCACGGCAGAAATTGAGTCTGCAATCCTAGACGACAGCATTGTAGCAGAGTGTGCGGTGGTGGGTTTCAATGATGACCTGACTGGACAGGCTGTGGCGGCCTTCGTAGTGTTGAAGAACAAGTCATCCTGGAGCACTGCCTCCGAATCCGAGTTACAGGACATTAAAAAACATTTGATTTTGGCGGTGAGAAAAGACATCGGACCCTTTGCAGCCCCCAAATTGATAGTACTGGTTGACGATCTACCCAAGACTCGTTCAGGTAAAATTATGAGGAGAATACTGCGTAAAATACTAGCTGGAGAAAGCGATCAATTAGGCGACGTCAGTACCTTATCAAATCCGGGCATTGTCCGTCACCTTATAGACTCAGTAAAGCTATAA\")\n",
    "oopt = Seq('atgagtccaagcaccgctccggcagttaaaccagacgctggtgactcatttggcataataggaaaactaatcaccatcttaaaagcaatggcctgctcaaagagatcccatgagattgttaagatggagcaaaaacaagaacagacgattgaatacgaacacctgattggaaagtctatagtaaagcagagatccatctccaattgattgtaatcagagttagctcaacactattcatcatacctgtcaggattacagaaatactagcaattatacttccaaagtttcgaatcatttaggatcttttttggcaactaagctggtcaattcttaaactggtttgaagacttcgacagcgtgtttgtatttaacccgtagacctgaatctttgcattagagaactatgcatggttcctaaatggcatcttaaatgcggcatactattgtgttgatcgacatgcattataaacgtccaactaaccttccatcatttatgaggttaacgtcacgtgtcagagatactcattaacatatgctgaattgttagaatagatttgtcaggtcgctcagatcttgcagtatagtttgggcgtttgtgagggagattcaatagaggtttatttgccaatgatttcgtaaacattgattgccttaatggcaatctccaggattggaattgtttactctgttgtgtttgttggtttttctagcaactgtctgcgttatcgaattgaccacaccaattccaggttagtctttgccaaggacgagtccaaacgtcggtgtaaaatccttgagaccaaaagaaagttctatgacgttgtaaaggaaacatcaagggtctgaaatgtacttgtatattagagaacaaatcattttaaggtgttcttggtctcagggcgtgatctggattggaacgaccagattgaaaaatataaaggatactgcacctgcaaatctttggatagcgtccactcgcttttcttattatactctacaagcattaagagttttgccaaaagcctttaatactgcacttcaagttatttgttaggttctcttctgaccatgtgttactcatagtactcccacaggcaagacgtatttttcacgtcccgtgatgtgggcaggattacttgacactccaactttgtttacgggtccatattatacggttgttcaacacttgtatagtaaggcattgccgccaatccaaactactttagttattgggatatcttagactaatataaagtcacgaaattttatgtggcttccaccgagattagattgttgaagaggtcttggcactttgacgtagagaacaactcattaaaatcattaaggtgtttaggtcctgtaggaatccccatctcagccaaagtgtggtactggtactgcatccaaatccgttaaaatgagattcttgtagtagatcttaattggcaaacaaagtttgggtgctatatgttaactcaggtagcaaggtgaattgcaccgttgaagtcagggtgcgcatgctttcccttctttggtttagaaccaattatattggatgagaactcgggaaaaaaaattgcagagacatatgcttaaagtctattggcgatgaaatccagatggtcctcctttgagagaacgatatggaaaaattatgatcggttcttagatacgtacatgaagtccatcaaagggtattacttctcaggagatggtgccaccagggattaagatggatatatgtggttcttgggcagggtcgacgacttcttaaatgtgtccagacactggatgtccaaggcaaaaattgtctctgcaatcttagactacagcattgtagcaatctgttagatggtgggtttctatgatgacctgacttgataggctgtggcggccatcttagtgttgaagaacttctcatccaggagcattaccaccaaatccgagttataggactttgaaaaacatttgattttggagatgagaaaagacatcggatccattgcagcccccaaattgatagtactggttgacgatttatccaagactggttcaagttaaattgtgaggtgaatactgtgtgaaatactagctggaaaaagcaatcaattaggcgacttcagtaccatatcaaattcgggcattgtctgttactttgtagactcaataatcatataa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGAGTCCAAGCACCGCTCCGGCAGTTAAACCAGACGCTGGTGACTCATCTAGCATAATAGGAGAACTAAAGTCCAAGCTAAAAGCAGTGGCCGGCTCAGAGAGATCCGATGAGGTTGCTGAGATGGAGCAAGAACAAGAACAGGCGACTGAATACGAACACCTGACTAGAGTCCCTATAGTAAAGCAGAGACCCATCTCCGATAGATTGCAACCAGAGTTAGCTCAACACTATTCACCACACCTGTCAGGATTACAGGAATACAAGCAATTATACAAGCAAAGTATCGAATCACCTGGGAAGTTTTTTGGCGACAAAGCTCGTCAATTCTTAAACTGGTTTAAAGACTTCGACAGCGTGTTTGTACCTAACCCGGAGACCGGAAAGCCTTCATTAGAGAACAATGCATGGTTCCTAAATGGCCAGTTAAATGCGGCATACAATTGTGTTGATAGACATGCACTACAAACGCCCAACAAACCTGCCATCATTTATGAGGCTGACGAGCCGGGTCAGGGATACACACTAACATATGCTGAATTGTTAGAACAGGTTTGTAAGGTCGCTCAGGTCTTGCAGTATAGTATGGGCGTTCGTAAGGGAGATACAGTAGCGGTTTATATGCCAATGATTCCGCAAACACTGATTACCTTAATGGCAATCACCAGGATTGGAGCTATTCACTCTGTTGTGTTTGCTGGTTTTTCTTGCAACAGTCTGCGTGATAGAATTAACGACGCCGATTCCAGGGTAGTCATTACCACGGACGAGTCCAAACGTGGGGGTAAAATCGTTGAGACCAAAAGAATCGTCGATGACGCTATAAAGGAAACATCAGGGGTCAGAAATGTACTTGTATATAAGAGAACAAATAATCCTAAGGTGCAGCTGGTCCCAGGGCGTGATCTGGATTGGGACGACGAGATTAAAAAATATAAAGGATACTGCCCCTGCGAACCTGTGGATAGCGAGCACCCGCTTTTCTTACTATACACTTCAGGCTCTACGGGTACTCCCAAAGGCGTTCAACACAGCACTGCAGGTTATTTGTTAGGTGCTCTTCTGACCATGCGTTACACATTCGACACCCACAGGGAAGACGTATTTTTCACGGCCGGTGATGTGGGCTGGATTACTGGACACACCTACGTTGTTTACGGGCCCCTATTATACGGTTGTACAACACTTGTATTCGAAGGCACTCCCGCCTATCCAAACTACTCTCGTTATTGGGATATCGTAGACAAATATAAAGTCACGCAATTTTATGTGGCTCCCACCGCGCTTAGATTGTTGAAGAGGGCTGGGGACTCTTACGTAGAGAACCACTCACTAAAATCACTAAGGTGTTTAGGTACTGTAGGAGAGCCCATCGCAGCCGAAGTGTGGGACTGGTACAGCGAGCAAATCGGTAAAAATGAGATTCCTATAGTAGATACTTATTGGCAAACAGAGTCTGGGAGCCATATGTTAACTCCGCTAGCAGGGGGAGTTACACCGATGAAGCCAGGGAGCGCATGCTTTCCCTTCTTTGGTATAGAACCAGCTATACTGGATGCGAACACGGGAAAAGAAATTACAGCGTCACATGCTGAAGGTGTACTGGCGGTGAAATCCGGATGGCCCTCCTTTGCGAGAACGATATGGAAAAATCATGATAGGTTCTTAGATACGTACCTGAAGCCCTTCAAAGGGTATTACTTCGCAGGAGATGGTGCCGCCAGGGATAAAGATGGATATATGTGGATCTTGGGCAGGGTCGACGACGTCGTAAATGTGTCCGGACACAGGCTGTCCACGGCAGAAATTGAGTCTGCAATCCTAGACGACAGCATTGTAGCAGAGTGTGCGGTGGTGGGTTTCAATGATGACCTGACTGGACAGGCTGTGGCGGCCTTCGTAGTGTTGAAGAACAAGTCATCCTGGAGCACTGCCTCCGAATCCGAGTTACAGGACATTAAAAAACATTTGATTTTGGCGGTGAGAAAAGACATCGGACCCTTTGCAGCCCCCAAATTGATAGTACTGGTTGACGATCTACCCAAGACTCGTTCAGGTAAAATTATGAGGAGAATACTGCGTAAAATACTAGCTGGAGAAAGCGATCAATTAGGCGACGTCAGTACCTTATCAAATCCGGGCATTGTCCGTCACCTTATAGACTCAGTAAAGCTATAA\n",
      "atgagtccaagcaccgctccggcagttaaaccagacgctggtgactcatttggcataataggaaaactaatcaccatcttaaaagcaatggcctgctcaaagagatcccatgagattgttaagatggagcaaaaacaagaacagacgattgaatacgaacacctgattggaaagtctatagtaaagcagagatccatctccaattgattgtaatcagagttagctcaacactattcatcatacctgtcaggattacagaaatactagcaattatacttccaaagtttcgaatcatttaggatcttttttggcaactaagctggtcaattcttaaactggtttgaagacttcgacagcgtgtttgtatttaacccgtagacctgaatctttgcattagagaactatgcatggttcctaaatggcatcttaaatgcggcatactattgtgttgatcgacatgcattataaacgtccaactaaccttccatcatttatgaggttaacgtcacgtgtcagagatactcattaacatatgctgaattgttagaatagatttgtcaggtcgctcagatcttgcagtatagtttgggcgtttgtgagggagattcaatagaggtttatttgccaatgatttcgtaaacattgattgccttaatggcaatctccaggattggaattgtttactctgttgtgtttgttggtttttctagcaactgtctgcgttatcgaattgaccacaccaattccaggttagtctttgccaaggacgagtccaaacgtcggtgtaaaatccttgagaccaaaagaaagttctatgacgttgtaaaggaaacatcaagggtctgaaatgtacttgtatattagagaacaaatcattttaaggtgttcttggtctcagggcgtgatctggattggaacgaccagattgaaaaatataaaggatactgcacctgcaaatctttggatagcgtccactcgcttttcttattatactctacaagcattaagagttttgccaaaagcctttaatactgcacttcaagttatttgttaggttctcttctgaccatgtgttactcatagtactcccacaggcaagacgtatttttcacgtcccgtgatgtgggcaggattacttgacactccaactttgtttacgggtccatattatacggttgttcaacacttgtatagtaaggcattgccgccaatccaaactactttagttattgggatatcttagactaatataaagtcacgaaattttatgtggcttccaccgagattagattgttgaagaggtcttggcactttgacgtagagaacaactcattaaaatcattaaggtgtttaggtcctgtaggaatccccatctcagccaaagtgtggtactggtactgcatccaaatccgttaaaatgagattcttgtagtagatcttaattggcaaacaaagtttgggtgctatatgttaactcaggtagcaaggtgaattgcaccgttgaagtcagggtgcgcatgctttcccttctttggtttagaaccaattatattggatgagaactcgggaaaaaaaattgcagagacatatgcttaaagtctattggcgatgaaatccagatggtcctcctttgagagaacgatatggaaaaattatgatcggttcttagatacgtacatgaagtccatcaaagggtattacttctcaggagatggtgccaccagggattaagatggatatatgtggttcttgggcagggtcgacgacttcttaaatgtgtccagacactggatgtccaaggcaaaaattgtctctgcaatcttagactacagcattgtagcaatctgttagatggtgggtttctatgatgacctgacttgataggctgtggcggccatcttagtgttgaagaacttctcatccaggagcattaccaccaaatccgagttataggactttgaaaaacatttgattttggagatgagaaaagacatcggatccattgcagcccccaaattgatagtactggttgacgatttatccaagactggttcaagttaaattgtgaggtgaatactgtgtgaaatactagctggaaaaagcaatcaattaggcgacttcagtaccatatcaaattcgggcattgtctgttactttgtagactcaataatcatataa\n"
     ]
    }
   ],
   "source": [
    "print(str(orig))\n",
    "print(str(oopt))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:dtailor27]",
   "language": "python",
   "name": "conda-env-dtailor27-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
